pkg/channelmetrics/observablechan.go

```diff
@@ -0,0 +1,84 @@
+// Package channelmetrics provides a flexible way to wrap Go channels with
+// additional metrics collection capabilities. This allows for monitoring
+// and tracking of channel usage and performance using different metrics backends.
+package channelmetrics
+
+import (
+	"time"
+
+	"github.com/trufflesecurity/trufflehog/v3/pkg/common"
+	"github.com/trufflesecurity/trufflehog/v3/pkg/context"
+)
+
+// MetricsCollector is an interface for collecting metrics. Implementations
+// of this interface can be used to record various channel metrics.
+type MetricsCollector interface {
+	RecordProduceDuration(duration time.Duration)
+	RecordConsumeDuration(duration time.Duration)
+	RecordChannelLen(size int)
+	RecordChannelCap(capacity int)
+}
+
+// ObservableChan wraps a Go channel and collects metrics about its usage.
+// It supports any type of channel and records metrics using a provided
+// MetricsCollector implementation.
+type ObservableChan[T any] struct {
+	ch        chan T
+	metrics   MetricsCollector
+	bufferCap int
+}
+
+// NewObservableChan creates a new ObservableChan wrapping the provided channel.
+// It records the channel's capacity immediately and sets up metrics collection
+// using the provided MetricsCollector and channel name. The chanName is used to
+// distinguish between metrics for different channels by incorporating it into
+// the metric names.
+func NewObservableChan[T any](ch chan T, metrics MetricsCollector) *ObservableChan[T] {
+	oChan := &ObservableChan[T]{
+		ch:        ch,
+		metrics:   metrics,
+		bufferCap: cap(ch),
+	}
+	oChan.RecordChannelCapacity() // Record capacity immediately
```

Comment:
Should we also record the length immediately? In most cases it'll be 0, but theoretically any channel can be converted into an `ObservableChan`.

Reply:
I didn't account for that use case, thanks.